### PR TITLE
KFSPTS-21168 creating customizations report bash script

### PR DESCRIPTION
**Pull request by jhulslander**
This is an initial start to the script.  Follow up user stories will add more functionality.

### CODE REVIEW

**Review by db483 — CHANGES_REQUESTED** [review]
The initial script looks good but I have some suggestions for improvement after the first round of review:

1. The readme file should:
- have instructions for running the script
- should explain what the script is used for
- what is the expected output
- summary of what to expect in the generated report/reports

2. I am not sure describing how it works behind the scene is needed.

3. Should the code also look at the KualiCo base code resources folders (We have C_X but no K_X)?

4. Does the script expect the developer running the script to update the boolean params in the script for each module to look at commit history and/or diff? It would be good to look for a way to update this so that these are passed in on the command line.

5. Technical report should be in a separate folder, for example ‘reports’.

6. Should the file name have a date/timestamp in it or have the reports placed in a timestamped sub folder? I expect we will be running this periodically and commit the reports.

7. I wonder if we should split the technical report into separate files by module so as to keep them more focused.

8. It may make sense to add a header summary with the number of files: overlays, extensions, xml?

9. Note that in the XML category there is xml and properties files; the xml ones are DD files, doc types, workflow, spring config, ojb. We should consider whether these should be reported is separate categories.

10. We may be able to somewhat simplify the code by dynamically determining the module names. 

11. We should separately report modules/code that is Cornell specific like rass, concur, pmw, I Want doc.

12. We should also generate a functional report. For now it could largely mimic the technical report (add commit history but not the diff) and then we can get some input from functionals as to what could be done to improve it and of course have a discussion as to what is feasible in an automated way. So perhaps the reports folder should have functional and technical sub folders.

13. The display of the diff at the end of the commit history could be improved by adding an empty line before. Also the header KualiCo file: … could be improved to say: “Diff with base code file” or “Diff between [base code file name] [cornell file name]”.

14. We may want to update the jira description to say "create initial script that generates report" so that it is clear this work will not be completed in this one user story.

Thank you.

**Comment by jhulslander** [comment]
Thanks for the feedback Daniela, Let me try re-organizing how this works

**Comment by jhulslander** [comment]
I've significantly reworked the script.  It processes one module at a time now, the various variables are controlled by command line input.  Also it generates  a report  with the module name, the report parameters, and a date stamp.  I think this script is cleaner, and more controllable.  This is ready for review again.

File: generate_report.sh [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+fi
+
+currentDate=`date`
+module="$1"
+commit_history="$2"
+kualico_diff="$3"
+
+echo "Processing module $module with commit history $commit_history and KualiCo diff $kualico_diff"
+
+reportFile="../reports/report_file_${module}"
```
Comment (db483) [inline generate_report.sh]: Would it be better to put this in a technical subfolder under reports?

File: generate_report.sh [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    kualicoJavaBase="${kualicoJavaBase}/kfs-core/"
+    kualiCoOverlayJavaBase="${kualiCoOverlayJavaBase}/${module}"
+    cornellJavaBase="${cornellJavaBase}/${module}"
+    cornellXMLBase="${cornellXMLBase}/${module}"
+    ;;
+esac
+
+KUALICO_JAVA=("KualiCo base overlay $module" "$kualiCoOverlayJavaBase" "$kualicoJavaBase" "$commit_history" "$kualico_diff")
+CORNELL_JAVA=("Cornell Java $module" "$cornellJavaBase" "" "$commit_history" "false")
+CORNELL_RESOURCES=("Cornell Resources $module" "$cornellXMLBase" "" "$commit_history" "false")
```
Comment (db483) [inline generate_report.sh]: Do we also have overlays for files in the resources folder?

File: generate_report.sh [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  moduleFileCount=$(find $modPath -type f -print | wc -l)
+  echo "The number of files in $modDescription : $moduleFileCount" >> $reportFile
+
+  moduleFiles=($(find $modPath -type f -follow -print))
+  for cuFile in "${moduleFiles[@]}"
+  do
+    echo "Cornell File $cuFile" >> $reportFile
+    
+    if [ "$showCommitHistory" = "true" ]; then
+      readarray -t gitLogEntries < <(git log --since=6.months --oneline -- $cuFile ":(exclude)phabricator")
```
Comment (db483) [inline generate_report.sh]: is it possible to remove the 6 month git history restriction?

**Review by db483 — CHANGES_REQUESTED** [review]
The script is much better and the output looks very good.

I just had a few things to consider for modification now and then we just need to make sure we capture all remaining changes in separate user stories.

Thank you.

**Comment by jhulslander** [comment]
I've added a technical and functional sub folder to the reports, moved the report into the technical folder.
Resource over lays are being procesed
The commit history now goes back to the beginning 
There is a pause so a specific branch or tagged can be checked out
The readme file now lists all the modules

this is ready for review again

File: README.md [hunk 4]
```diff
@@ -1,3 +1,8 @@
 # cu-kfs-customizations-reports
-Bash scripts to generate reports showing customizations we have made to our KFS overlay.
-To run this job, use bash.
+This project contains a bash script that analyzes CU-KFS code, shows recent commit history per file, and when it is an overylay of a basecode file, it shows the differences between the CU-KFS code on the KualiCo code.
+
+# How to run this script
+bash generate_report.sh fp true true
```
Comment (db483) [inline README.md]: Should the readme file also explain how branches can be switched when the script pauses and what the default branches are?

**Review by db483 — CHANGES_REQUESTED** [review]
The changes look good. I had a question about whether the readme file should contain details about switching the branches.

Since this did not implement all suggestions for improvement should the user story be updated to specify what has been implemented, what is still left to be implemented and then create user stories for next steps? Thank you.

File: generate_report.sh (hunk 3 above)
Comment (jhulslander) [reply generate_report.sh]: Yes, it is pulling the full history now

File: generate_report.sh (hunk 2 above)
Comment (jhulslander) [reply generate_report.sh]: yes, added

File: generate_report.sh (hunk 1 above)
Comment (jhulslander) [reply generate_report.sh]: sure, added a technical and functional folder, and moved the report to the technical folder

File: README.md (hunk 4 above)
Comment (jhulslander) [reply README.md]: sure, can add note

**Comment by jhulslander** [comment]
I just added a summary seciont
reviewing your notes above, I think the un-implemented suggestions are 
1. Note that in the XML category there is xml and properties files; the xml ones are DD files, doc types, workflow, spring config, ojb. We should consider whether these should be reported is separate categories.
2. We should also generate a functional report. For now it could largely mimic the technical report (add commit history but not the diff) and then we can get some input from functionals as to what could be done to improve it and of course have a discussion as to what is feasible in an automated way. So perhaps the reports folder should have functional and technical sub folders.

for number 1, I renamed the section to resources rather than XML. If resources should be broken out further, then perhaps that can be a follow up user story
for number 2, I think we need a user story to identify what the functional report should have, and then implement that report

**Comment by jhulslander** [comment]
This is ready for review again

**Review by dp65 — COMMENTED** [review]
Would it be better if the command line arguments were changed to options from positional parameters?  

- It allows either of the boolean parameters to be specified independently.
- It would be self-documenting, without the need to refer to the README.  I would want to call the script like one of these:
- - generate_report.sh --module=sys --show-commit-history=false --show-kualico-diff=true
- - generate_report.sh -m sys -d true -c false
- - generate_report.sh -h (to show the options and the valid values for each).
- It doesn't add too much code or complexity. I have similar code you can steal in the https://github.com/CU-CommunityApps/batch-scripts/blob/dev-aws/batch.sh script.

**Comment by db483** [comment]
@dp65 I like your suggestion. @jhulslander Should we add an improvement user story for this?

**Comment by jhulslander** [comment]
let me try and switch to positional arguments

**Comment by db483** [comment]
Thank you for summarizing Jay. I think some parts of #9, #10 and #12 on my initial list still need some work like you noted. 
For #9 it may be helpful to separate ojb, workflow, jsp, files in separate sections.
For #10 you are now generating things by module but that may not address the initial requirement from the functionals to automatically report local implementations like the integrations to pmw, concur, and I think the trickier ones like the I Want document. Not sure how that can be done.
For #12 like you noted, the functional report needs to be added.

Additionally, please consider adding a user story for Donivan's suggestion.

It would be good to add our conclusions on the KFSPTS-21168 user story. I will then merge this.

**Comment by jhulslander** [comment]
I have adjusted the input as Donivan had suggested

**Comment by jhulslander** [comment]
I added what I think is a descent summary of things not done yet.

This is ready for review again

**Review by dp65 — COMMENTED** [review]
@jhulslander, thanks for changing it to parameters. 
Lines 14-17 appear to be extra duplicates of lines 10-13.
Other than that, it looks good.

File: generate_report.sh [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+  --show-commit-history)
+    shift
+    commit_history="$1"
+    ;;
+  --show-kualico-diff)
+    shift
+    kualico_diff="$1"
+    ;;
+  * )
+    echo "Examples of running this script are"
```
Comment (db483) [inline generate_report.sh]: Should this be displayed when just bash generate_report.sh is run?  I tried running 'bash generate_report.sh' and I get 'invalid module entered'.

**Review by db483 — COMMENTED** [review]
Jay,

The updated script looks great.

I noticed the summary section may need an update as I believe it is printing an 'n' instead of a new line:
******************************** Summary Section *****************************************
 n KualiCo base Java overlay sys :       13 n KualiCo base resources overlay sys :        0 n Cornell Java sys :      117 n Cornell Resources sys :       24
******************************************************************************************

I also get an 'invalid module entered' message when I ran just bash generate_report.sh, I expected to get the more detailed message about how to run the script.

These are minor issues and I am happy with them being addressed in a future story if that is ok with you.

Thank you.

File: generate_report.sh (hunk 5 above)
Comment (jhulslander) [reply generate_report.sh]: I have fixed the help message logic, it will print when no params at all are entered now

**Comment by jhulslander** [comment]
The summary section is printing correct for me
`******************************** Summary Section *****************************************
 
 KualiCo base Java overlay ld :        3 
 KualiCo base resources overlay ld :        0 
 Cornell Java ld :       27 
 Cornell Resources ld :       19
******************************************************************************************`

I viewed the report in atom and testEdit, both looked fine.  I think this could be handled in a follow up story

**Review by db483 — APPROVED** [review]
Changes look good. bash generate_report.sh now shows the expected message.
I still cannot see the new lines in the summary section but that could be because of a difference in the bash version. If validation will show an issue for other developers too we can open a follow up user story to look into this. Merging.

**Pull request merged**